### ISSUE TITLE
About format settings of validation dataset

### ISSUE THREAD

**Issue by daegonYu** [UKPLab/sentence-transformers#2740]
In the example code below, the format of eval_dataset is a format available for both loss and TripletEvaluator. My error occurred when I applied InformationRetrievalEvaluator instead of TripletEvaluator. How do I apply InformationRetrievalEvaluator during learning? I am attaching part of the training code at the bottom.

In short, I want to keep the loss function CachedMultipleNegativesRankingLoss or MultipleNegativesRankingLoss and set the evaluator to InformationRetrievalEvaluator during training. I would appreciate your help on this.

Below is training example code from SBERT.net
```py

from datasets import load_dataset
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
    SentenceTransformerModelCardData,
)
from sentence_transformers.losses import MultipleNegativesRankingLoss
from sentence_transformers.training_args import BatchSamplers
from sentence_transformers.evaluation import TripletEvaluator

# 1. Load a model to finetune with 2. (Optional) model card data
model = SentenceTransformer(
    "microsoft/mpnet-base",
    model_card_data=SentenceTransformerModelCardData(
        language="en",
        license="apache-2.0",
        model_name="MPNet base trained on AllNLI triplets",
    )
)

# 3. Load a dataset to finetune on
dataset = load_dataset("sentence-transformers/all-nli", "triplet")
train_dataset = dataset["train"].select(range(100_000))
eval_dataset = dataset["dev"]
test_dataset = dataset["test"]

# 4. Define a loss function
loss = MultipleNegativesRankingLoss(model)

# 5. (Optional) Specify training arguments
args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/mpnet-base-all-nli-triplet",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    warmup_ratio=0.1,
    fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=False,  # Set to True if you have a GPU that supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
    # We can use ROUND_ROBIN or PROPORTIONAL - to avoid focusing too much on one dataset, we will
    # use round robin, which samples the same amount of batches from each dataset, until one dataset is empty
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=100,
    save_strategy="steps",
    save_steps=100,
    save_total_limit=2,
    logging_steps=100,
    run_name="mpnet-base-all-nli-triplet",  # Will be used in W&B if `wandb` is installed
)

# 6. (Optional) Create an evaluator & evaluate the base model
dev_evaluator = TripletEvaluator(
    anchors=eval_dataset["anchor"],
    positives=eval_dataset["positive"],
    negatives=eval_dataset["negative"],
    name="all-nli-dev",
)
dev_evaluator(model)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss=loss,
    evaluator=dev_evaluator,
)
trainer.train()

# (Optional) Evaluate the trained model on the test set
test_evaluator = TripletEvaluator(
    anchors=test_dataset["anchor"],
    positives=test_dataset["positive"],
    negatives=test_dataset["negative"],
    name="all-nli-test",
)
test_evaluator(model)

# 8. Save the trained model
model.save_pretrained("models/mpnet-base-all-nli-triplet/final")

# 9. (Optional) Push it to the Hugging Face Hub
model.push_to_hub("mpnet-base-all-nli-triplet")

```

<br>
My training code

```py

from datasets import load_dataset
import random

corpus = load_dataset("BeIR/quora", "corpus", split="corpus")
queries = load_dataset("BeIR/quora", "queries", split="queries")
relevant_docs_data = load_dataset("BeIR/quora-qrels", split="validation")

# Shrink the corpus size heavily to only the relevant documents + 10,000 random documents
required_corpus_ids = list(map(str, relevant_docs_data["corpus-id"]))
required_corpus_ids += random.sample(corpus["_id"], k=10_000)
corpus = corpus.filter(lambda x: x["_id"] in required_corpus_ids)

# Convert the datasets to dictionaries
corpus = dict(zip(corpus["_id"], corpus["text"]))  # Our corpus (cid => document)
queries = dict(zip(queries["_id"], queries["text"]))  # Our queries (qid => question)
relevant_docs = {}  # Query ID to relevant documents (qid => set([relevant_cids])
for qid, corpus_ids in zip(relevant_docs_data["query-id"], relevant_docs_data["corpus-id"]):
    qid = str(qid)
    corpus_ids = str(corpus_ids)
    if qid not in relevant_docs:
        relevant_docs[qid] = set()
    relevant_docs[qid].add(corpus_ids)


valid_dataset = {}
valid_dataset['corpus'] = corpus
valid_dataset['queries'] = queries
valid_dataset['relevant_docs'] = relevant_docs

evaluator = InformationRetrievalEvaluator(
    queries=valid_dataset['queries'],
    corpus=valid_dataset['corpus'],
    relevant_docs=valid_dataset['relevant_docs'],
    name="dev",
)

model = ...
args = ...
train_dataset  = ...  # (anchor, positive) pairs
loss = MultipleNegativesRankingLoss(model)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=valid_dataset,
    loss=loss,
    evaluator=evaluator,
)
trainer.train()

```

**Comment by tomaarsen** [comment]
Hello!

The `eval_dataset` passed to the trainer should have the same format as the `train_dataset`, as it'll indeed be used with the loss function. And as you also noticed, the inputs for the IR Evaluator is different. 
My recommendation is to make the corpus/query/relevant_docs variables as variables that are separate from the eventual `eval_dataset` variable. So, exactly like this:

https://huggingface.co/tomaarsen/mpnet-base-gooaq/blob/main/train_script.py#L62-L89

I hope that helps!

- Tom Aarsen

**Comment by daegonYu** [comment]
I completely understand. thank you!
